### PR TITLE
character set => move logic into parser

### PR DESCRIPTION
**Pull request by ColinEberhardt**

### CODE REVIEW

File: assembly/nfa/matcher.ts [hunk 1]
```diff
@@ -79,28 +112,20 @@ export class CharacterClassMatcher extends Matcher {
   }
 }
 
+// no closure support
+let _code: u32;
+
 export class CharacterSetMatcher extends Matcher {
-  constructor(public set: string, public negated: bool) {
+  constructor(public matchers: Matcher[], public negated: bool) {
```
Comment (ColinEberhardt) [inline assembly/nfa/matcher.ts:118]: @MaxGraey - I've hit a funny error here, wondering if you might know the root cause?

~~~
  RuntimeError: unreachable
        at assembly/nfa/matcher/Matcher#matches@virtual (wasm-function[240]:1)
        at assembly/nfa/matcher/CharacterSetMatcher#matches~anonymous|0 (wasm-function[241]:19)
        at ~lib/array/Array<assembly/nfa/matcher/Matcher>#some (wasm-function[242]:85)
        at assembly/nfa/matcher/CharacterSetMatcher#matches (wasm-function[244]:21)
        at assembly/nfa/nfa/MatcherState<assembly/nfa/matcher/CharacterSetMatcher>#matches (wasm-function[245]:8)
        at assembly/nfa/nfa/State#matches@virtual (wasm-function[250]:58)
        at assembly/regexp/recursiveBacktrackingSearch (wasm-function[184]:121)
        at assembly/regexp/recursiveBacktrackingSearch@varargs (wasm-function[185]:56)
        at assembly/regexp/RegExp#exec (wasm-function[192]:307)
~~~

works just fine in TS/JS

Comment (MaxGraey) [reply assembly/nfa/matcher.ts:118]: Hmm, hard to tell. Need minimal repro

Comment (MaxGraey) [reply assembly/nfa/matcher.ts:118]: I tried reproduce this issue in minimal example but can't:
[playground](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1ub25lJnJ1bnRpbWU9aGFsZgpjbGFzcyBNYXRjaGVyIHsKICBtYXRjaGVzKGNvZGU6IHUzMik6IGJvb2wgewogICAgcmV0dXJuIGZhbHNlOwogIH0KfQoKY2xhc3MgQ2hhcmFjdGVyUmFuZ2VOb2RlIGV4dGVuZHMgTWF0Y2hlciB7CiAgY29uc3RydWN0b3IoKSB7IHN1cGVyKCk7IH0KCiAgbWF0Y2hlcyhjb2RlOiB1MzIpOiBib29sIHsKICAgIHJldHVybiB0cnVlOwogIH0KfQoKY2xhc3MgQ2hhcmFjdGVyTm9kZSBleHRlbmRzIE1hdGNoZXIgewogIGNvbnN0cnVjdG9yKCkgeyBzdXBlcigpOyB9CgogIG1hdGNoZXMoY29kZTogdTMyKTogYm9vbCB7CiAgICByZXR1cm4gZmFsc2U7CiAgfQp9CgpsZXQgX2NvZGU6IHUzMjsKCmNsYXNzIENoYXJhY3RlclNldE1hdGNoZXIgZXh0ZW5kcyBNYXRjaGVyIHsKICBjb25zdHJ1Y3RvcihwdWJsaWMgbWF0Y2hlcnM6IE1hdGNoZXJbXSwgcHVibGljIG5lZ2F0ZWQ6IGJvb2wgPSBmYWxzZSkgewogICAgc3VwZXIoKTsKICB9CgogIG1hdGNoZXMoY29kZTogdTMyKTogYm9vbCB7CiAgICBfY29kZSA9IGNvZGU7CiAgICBpZiAoIXRoaXMubmVnYXRlZCkgewogICAgICByZXR1cm4gdGhpcy5tYXRjaGVycy5zb21lKChtKSA9PiBtLm1hdGNoZXMoX2NvZGUpKTsKICAgIH0gZWxzZSB7CiAgICAgIHJldHVybiAhdGhpcy5tYXRjaGVycy5zb21lKChtKSA9PiBtLm1hdGNoZXMoX2NvZGUpKTsKICAgIH0KICB9Cn0KCmV4cG9ydCBmdW5jdGlvbiB0ZXN0KCk6IGJvb2wgewogIGNvbnN0IG1hdGNoZXJzID0gbmV3IEFycmF5PE1hdGNoZXI+KCk7CiAgbWF0Y2hlcnMucHVzaChuZXcgQ2hhcmFjdGVyUmFuZ2VOb2RlKTsKICBtYXRjaGVycy5wdXNoKG5ldyBDaGFyYWN0ZXJOb2RlKTsKICByZXR1cm4gbmV3IENoYXJhY3RlclNldE1hdGNoZXIobWF0Y2hlcnMsIHRydWUpLm1hdGNoZXMoMCk7Cn0KIyFodG1sCjx0ZXh0YXJlYSBpZD0ib3V0cHV0IiBzdHlsZT0iaGVpZ2h0OiAxMDAlOyB3aWR0aDogMTAwJSIgcmVhZG9ubHk+PC90ZXh0YXJlYT4KPHNjcmlwdD4KbG9hZGVyLmluc3RhbnRpYXRlKG1vZHVsZV93YXNtLCB7IC8qIGltcG9ydHMgKi8gfSkKICAudGhlbigoeyBleHBvcnRzIH0pID0+IHsKICAgIGNvbnN0IG91dHB1dCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdvdXRwdXQnKQogICAgb3V0cHV0LnZhbHVlICs9ICdwdHI6ICcgKyBleHBvcnRzLnRlc3QoKTsKICB9KQo8L3NjcmlwdD4=)

Comment (ColinEberhardt) [reply assembly/nfa/matcher.ts:118]: Unfortunately I think finding a minimal reproduction could be a challenge. I tried a similar approach, but couldn't reproduce.

What I have found is that adding the following block of code, executed before the RegExp is created, removes the error:

https://github.com/ColinEberhardt/assemblyscript-regex/pull/12/commits/1592687c388bcd46d678d022bcd842149855a94e#diff-f350832f4859b9724e3ef59f756d2d201226fd81fd797207367fc269581115adR171-R179

Comment (ColinEberhardt) [reply assembly/nfa/matcher.ts:118]: I've raised an issue here https://github.com/AssemblyScript/assemblyscript/issues/1635

**Pull request merged**